hierarqcal/utils.py

```diff
@@ -205,4 +205,96 @@ def plot_motif(
 #             # oPlot.add_plot(ax)
 #             figs.append(fig)
 #             plt.close()
-#     return figs
\ No newline at end of file
+#     return figs
+
+
+def get_circ_info_from_string(input_str):
```

Comment:
The code is very robust and working well! The logic is sound.

I have a suggestion around the organisation of the functions:
I think this function should be part of `Qunitary`, both this code and Qunitary's intialization calculates `unique_params` and `unique_bits`. Since getting the `substr_list` is backend independant it can form part of Qunitary and run when initialized. During initialization store substr_list as an attribute of Qunitary (which by default should be None unless a string was provided). With substr_list as an attribute for Qunitary, I think a good name would be circuit_instructions and we should change it's type to a namedTuple (then we can reference the instructions by key) i.e. circuit_instructions.gate_name, circuit_instructions.symbol_info, circuit_instructions.sub_bits

So what I'm thinking about looks something along the lines:
```python
# core.py
from collections import namedtuple
# ...
CircuitInstructions = namedtuple('CircuitInstructions', ['gate_name', 'symbol_info', 'sub_bits'])
# ...
class Qunitary():
     def __init__
          self.circuit_instructions = None
          if isinstance(function, str):
                circuit_instructions, unique_params, unique_bits = self.get_circuit_from_string(function)
                # Do the self.attribute = stuff 


```
Then the get_circuit_from_string in qiskit helper stays the same, only it gets circuit instructions directly from the qunitary object sent to it, i.e. `qunitary.circuit_instructions.gate_name` etc

Reply:
Looking at this again, I realize it should rather be:
```python
CircuitInstruction= namedtuple('CircuitInstruction', ['gate_name', 'symbol_info', 'sub_bits'])
```
So that every circuit_instruction is a named tuple, then the Qunitary has the attribute circuit_instructions which is a list `[CircuitInstruction(...), CircuitInstruction(...)]`

Reply:
ah I see, this is nice! implemented